### PR TITLE
Support for sparse data structures in tidymodels

### PR DESCRIPTION
**Pull request by juliasilge**
This blog post should be published **after** the main tune blog post announcement by at least a day or so, and is flexible about when exactly it goes out.

### CODE REVIEW

**Comment by juliasilge** [comment]
I can't assign reviewers in this repo but @topepo take a look! 👀

File: content/blog/tidymodels-sparse-support/index.Rmd [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+```{r setup, include=FALSE}
+library(knitr)
+knitr::opts_chunk$set(cache = TRUE, cache.lazy = FALSE, warning = FALSE, 
+                      message = FALSE, echo = TRUE, dpi = 180,
+                      fig.width = 8, fig.height = 5)
+library(tidyverse)
+theme_set(theme_minimal())
+```
+
+The new release of [tune](linktopost) is chock full of improvements and new features. When combined with the latest releases of [hardhat](http://hardhat.tidymodels.org/) and [parsnip](https://parsnip.tidymodels.org/), one upgrade that tidymodels users can now use in their day-to-day modeling work is **support for sparse data structures**.
```
Comment (topepo) [inline content/blog/tidymodels-sparse-support/index.Rmd:29]: "limited support" maybe? I'm worried about people starting with a sparse matrix

File: content/blog/tidymodels-sparse-support/index.Rmd [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+  count(product, word) %>%
+  bind_tf_idf(word, product, n) %>%
+  cast_dfm(product, word, tf_idf)
+```
+
+As is typical for text data, this document-feature matrix is extremely sparse, with many zeroes. Most documents do not contain most words. If we coerce this kind of structure to anything like a `data.frame`, we run into two challenges:
+
+- The amount of **memory** this object requires increases dramatically.
+- We lose out on the **speed** gained from any specialized model algorithms built for sparse data.
+
```
Comment (topepo) [inline content/blog/tidymodels-sparse-support/index.Rmd:59]: I'd briefly mention how sparse matrices work under the hood and maybe do an `obj_size()` calculation with it and its data frame version.

File: content/blog/tidymodels-sparse-support/index.Rmd [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+## A blueprint for sparse models
+
+Before the most recent releases of hardhat, parsnip, and tune, there was no support for sparse data structures within tidymodels. Now, you can specify a hardhat **blueprint** for sparse data.
+
+```{r}
+library(hardhat)
+sparse_bp <- default_recipe_blueprint(composition = "dgCMatrix")
+```
+
+The `dgCMatrix` composition is from the [Matrix](https://cran.r-project.org/package=Matrix) package, and is the most standard class for sparse numeric matrices in modeling in R. You can also specify a dense matrix composition with `composition = "matrix"`.
```
Comment (topepo) [inline content/blog/tidymodels-sparse-support/index.Rmd:69]: Somewhere, maybe here, we should differentiate what can _can_ do now versus what you _still can't_ do. IIRC you can start with a sparse matrix only if you use `fit_xy()` with a `parsnip` model. For example: 

```r
library(tidymodels)
library(Matrix)
library(hardhat)

data("hpc_data", package = "modeldata")

# ------------------------------------------------------------------------------

# Can do this: 
new_x <- Matrix::sparse.model.matrix(class ~ protocol + day, data = hpc_data)

cls_spec <- 
  multinom_reg(penalty = .1, mixture = 0.5) %>% 
  set_engine("glmnet", nlambda = 5)

# input is sparse
fit_xy(cls_spec, x = new_x, y = hpc_data$class)

# ------------------------------------------------------------------------------
# can do this:

sparse_bp <- default_recipe_blueprint(composition = "dgCMatrix")

rec <- 
  recipe(class ~ protocol + day, data = hpc_data) %>% 
  step_dummy(all_predictors())

wflow <- 
  workflow() %>% 
  add_recipe(rec, blueprint = sparse_bp) %>% 
  add_model(cls_spec)

#input still has to be a data frame but data gets converted and passed to model as sparse
wflow %>% fit(data = hpc_data)

# can't do anything else really
```

Somewhere we should say this.

File: content/blog/tidymodels-sparse-support/index.Rmd [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+  workflow() %>%
+  add_recipe(text_rec, blueprint = sparse_bp) %>%
+  add_model(lasso_spec)
+  
+wf_default <- 
+  workflow() %>%
+  add_recipe(text_rec) %>%
+  add_model(lasso_spec)
+```
+
```
Comment (topepo) [inline content/blog/tidymodels-sparse-support/index.Rmd:109]: Start a new subsection on modeling

File: content/blog/tidymodels-sparse-support/index.Rmd [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+fit_resamples(wf_default, food_folds) %>%
+  collect_metrics()
+```
+
+To see a detailed text modeling example using this dataset of food reviews, _without_ sparse encodings but complete with tuning hyperparameters, check out [our article on `tidymodels.org`](https://www.tidymodels.org/learn/work/tune-text/)!
+
+## Current limits
+
+In tidymodels, the support for sparse data structures begins coming _out_ of a [preprocessing recipe](https://www.tmwr.org/recipes.html) and continues throughout the fitting and tuning process. There is heterogeneity in how recipes themselves handle data internally; this is why we didn't a huge decrease in memory use when comparing `wf_sparse` to `wf_default`. The [textrecipes](https://textrecipes.tidymodels.org/) package internally adopts the idea of a [tokenlist](https://textrecipes.tidymodels.org/reference/tokenlist.html), which is memory efficient for sparse data, but other recipe steps may handle data in a dense tibble structure. Keep these current limits in mind as you consider the memory requirements of your modeling projects!
+
```
Comment (topepo) [inline content/blog/tidymodels-sparse-support/index.Rmd:146]: list the three models that support a "sparse handoff"

File: content/blog/tidymodels-sparse-support/index.Rmd (hunk 1 above)
Comment (topepo) [inline content/blog/tidymodels-sparse-support/index.Rmd:29]: Mention that this is the second of three articles related to tune 0.1.2 and use link to previous (which should be `/blog/2020/11/tune-0-1-2/`)

**Review by topepo — CHANGES_REQUESTED** [review]

**Comment by topepo** [comment]
One other question... should we drop all three posts the same day? @batpigandme does this matter?

**Comment by juliasilge** [comment]
Another option would be to post the main tune announcement one day and the followup posts both the day after? I don't hold any strong opinions on this.

**Comment by batpigandme** [comment]
> One other question... should we drop all three posts the same day? @batpigandme does this matter?

While it doesn't technically matter, I think it can sort of "dilute" the impact—I don't think there'd be any harm in spacing them out across days (though feel free to override, of course).

**Comment by topepo** [comment]
`tune` was accepted this morning. I'll post the first article on Monday (binaries might be built by then). Let's post this article next Wednesday and I'll post the third on Friday.

**Pull request merged**